src/main/java/org/opensearch/neuralsearch/processor/NormalizationProcessorWorkflow.java

```diff
@@ -202,20 +207,42 @@ private void updateOriginalQueryResults(final CombineScoresDto combineScoresDTO,
         final List<CompoundTopDocs> queryTopDocs = getCompoundTopDocs(combineScoresDTO, querySearchResults);
         final Sort sort = combineScoresDTO.getSort();
         int totalScoreDocsCount = 0;
-        for (int index = 0; index < querySearchResults.size(); index++) {
-            QuerySearchResult querySearchResult = querySearchResults.get(index);
-            CompoundTopDocs updatedTopDocs = queryTopDocs.get(index);
-            totalScoreDocsCount += updatedTopDocs.getScoreDocs().size();
-            TopDocsAndMaxScore updatedTopDocsAndMaxScore = new TopDocsAndMaxScore(
-                buildTopDocs(updatedTopDocs, sort),
-                maxScoreForShard(updatedTopDocs, sort != null)
+
+        // Get index of first non-empty CompoundTopDocs to check if collapse is enabled
+        boolean isCollapseEnabled = false;
+        int indexOfFirstNonEmpty = 0;
+        for (CompoundTopDocs compoundTopDocs : queryTopDocs) {
+            if (!compoundTopDocs.getTopDocs().isEmpty() && compoundTopDocs.getTopDocs().getFirst() instanceof CollapseTopFieldDocs) {
+                isCollapseEnabled = true;
+                break;
+            }
+            indexOfFirstNonEmpty++;
+        }
+        if (isCollapseEnabled) {
+            totalScoreDocsCount = collapse(
+                queryTopDocs,
+                indexOfFirstNonEmpty,
+                querySearchResults,
+                sort,
+                isFetchPhaseExecuted,
+                combineScoresDTO
             );
-            // Fetch Phase had ran before the normalization phase, therefore update the from value in result of each shard.
-            // This will ensure the trimming of the search results.
-            if (isFetchPhaseExecuted) {
-                querySearchResult.from(combineScoresDTO.getFromValueForSingleShard());
+        } else {
+            for (int shardIndex = 0; shardIndex < querySearchResults.size(); shardIndex++) {
```

Comment:
Can this piece of code and the code used later in updateOriginalQuerySearchResultWithCollapseEnabled has similar structure. Can we combine them?

Reply:
I refactored this part into a separate class in my last commit